src/data/nat/hyperoperation.lean

```diff
@@ -0,0 +1,109 @@
+/-
+Copyright (c) 2023 Mark Andrew Gerads. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mark Andrew Gerads, Junyan Xu, Alex J. Best
+-/
+import tactic.ring
+
+/-!
+# Hyperoperation sequence
+
+This file defines the Hyperoperation sequence.
+`hyperoperation 0 m k = k + 1`
+`hyperoperation 1 m k = m + k`
+`hyperoperation 2 m k = m * k`
+`hyperoperation 3 m k = m ^ k`
+`hyperoperation (n + 1) m (k + 1) = hyperoperation n m (hyperoperation (n + 1) m k)`
+
+## References
+
+* <https://en.wikipedia.org/wiki/Hyperoperation>
+
+## Tags
+
+hyperoperation
+-/
+
+/--
+Implementation of the hyperoperation sequence
+-/
+def hyperoperation : ℕ → ℕ → ℕ → ℕ
```

Comment:
Does this make sense to generalize to  something like
```suggestion
def hyperoperation : ℕ → ℕ → α → ℕ
```

Reply:
I don't know how the generalization would work; if this can be generalized, someone else can do it. As I understand it, there is https://en.wikipedia.org/wiki/Tetration a holomorphic solution to `hyperoperation 4 (exp 1) : ℂ → ℂ`, but I don't know about `hyperoperation 4 : ℂ → ℂ → ℂ` in general, let alone `hyperoperation : ℕ → ℂ → ℂ → ℂ`.

Reply:
I just tried this, and you're right; it's not as easy as I thought.